src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Net.Sockets
+{
+    // TODO:
+    // - ReceiveMessageFrom
+    // - Plumb status through async APIs to avoid callbacks on synchronous completion
+    //     - NOTE: this will require refactoring in the *Async APIs to accommodate the lack
+    //             of completion posting
+    // - Add support for unregistering + reregistering for events
+    //     - This will require a new state for each queue, unregistred, to track whether or
+    //       not the queue is currently registered to receive events
+
+    sealed class SocketAsyncContext
+    {
+        private abstract class AsyncOperation
+        {
+            public AsyncOperation Next;
+            public SocketError ErrorCode;
+
+            public AsyncOperation()
+            {
+                Next = this;
+            }
+
+            public abstract void Complete();
+        }
+
+        private sealed class SentinelAsyncOperation : AsyncOperation
+        {
+            public override void Complete()
+            {
+                Debug.Fail("SentinelAsyncOperation.Complete() should never be called");
+            }
+        }
+
+        private sealed class TransferOperation : AsyncOperation
+        {
+            public Action<int, byte[], int, int, SocketError> Callback;
+            public byte[] Buffer;
+            public BufferList Buffers;
+            public int BufferIndex;
+            public int Offset;
+            public int Count;
+            public int Flags;
+            public byte[] SocketAddress;
+            public int SocketAddressLen;
+            public int BytesTransferred;
+            public int ReceivedFlags;
+
+            public override void Complete()
+            {
+                Debug.Assert(Callback != null);
+
+                Callback(BytesTransferred, SocketAddress, SocketAddressLen, ReceivedFlags, ErrorCode);
+            }
+        }
+
+        private abstract class AcceptOrConnectOperation : AsyncOperation
+        {
+            public byte[] SocketAddress;
+            public int SocketAddressLen;
```

Comment:
I see that these two fields also exist in TransferOperation.  Any value in having TransferOperation derive from this type (with a different name)?